README
Visio version ___EUMM_VERSION___
=============

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2012 by Thomas McMeekin

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Visio.

Considerations	tomby	Mon 18 Sep 2023 07:02:25 AEST

The Visio format changed from straight XML to the OPC convention c. 2013.
There are some sample scripts on use of OPC in the example folder, e.g. vsdx.pl

Other references below:

  https://en.wikipedia.org/wiki/Open_Packaging_Conventions

  https://learn.microsoft.com/en-us/office/client-developer/visio/introduction-to-the-visio-file-formatvsdx

  https://www.ecma-international.org/publications-and-standards/standards/ecma-376/
  https://metacpan.org/pod/OPC#AUTHOR

Suggested approach:
a. Create a new module, e.g.Visio::OPC which handles the interaction with the 
   OPC-format file.  This would handle the read and save operations via the OPC
   module.
b. The OPC format is basically a collection of parts/relations in the form of
   XML fragments.  So LibXML is still fundamental to reading this format, e.g.
   OPC indexes all the document parts and fragments, but you need XML to pull
   the XML schemas out for the relationship IDs.
c. Retain the Visio-XML for basic operations? Or delegate the XML components.

TO-DO.

Visio Components are the items that make up the document structure.
page create
page delete
page properties
document properties

Visio Actions allow shapes to be manipulated
master (shape) add
page (shape) add
copy shape master->page
copy shape page->page (within, between)

shape properties, e.g. name, position, size
attribute assignment

0.01  Sun Jun  3 14:12:44 2012
	- original version; created by h2xs 1.23 with options
		-X -n Visio



Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my $preop = 'find $(DISTVNAME) -type f -exec egrep -q ___EUMM_VERSION___ {} \; -exec sed -i "s/___EUMM_VERSION___/Build V$(VERSION)/" {} \; -print ;' .
'perldoc -uT $(DISTVNAME)/$(VERSION_FROM) > README.pod';

WriteMakefile(
    NAME              => 'Visio',
    VERSION_FROM      => 'lib/Visio.pm', # finds $VERSION
    MIN_PERL_VERSION  => '5.006004',
    LICENSE           => 'gpl_2',
    PREREQ_PM         => {
	'Visio::XML' => '1.001',
    }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Visio.pm', # retrieve abstract from module
       AUTHOR         => 'tmcmeeki <tmcmeeki@cpan.org>') : ()),
        dist => { PREOP => $preop, },
);


lib/Visio.pm
package Visio;
#
# Visio.pm - revised class for Visio.
# Revision ___EUMM_VERSION___, Copyright (C) 2012 Thomas McMeekin
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License,
# or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#

=head1 NAME

Visio - revised class for Visio.

=head1 SYNOPSIS

	use Visio;


=head1 DESCRIPTION

___detailed_class_description_here___

=over 4

=item 1.  tba

tba

=back

=cut

use strict;

use vars qw/ $VERSION /;
#use vars qw/ @ISA $VERSION /;

use Carp qw(cluck confess); # only use stack backtrace within class
use Data::Dumper;

use File::Basename;
use POSIX;
use Visio::XML;
use Logfer qw/ :all /;


# package constants

use constant ATTR_SEARCH => qw/ BaseID ID UniqueID NameU /;	# list of attributes to be returned by a given search

use constant EXT_DOCUMENT  => ".vdx";

use constant NN_DOCUMENT => "VisioDocument";

use constant NN_SECTIONS => qw/ DocumentProperties DocumentSettings Colors
  FaceNames StyleSheets DocumentSheet Masters Pages Windows /;


# package globals
our $AUTOLOAD;
$VERSION = "1.001";     # update this on new release
#@ISA = ( "XML" );
#@EXPORT = qw();


# package locals
my $_n_objects = 0;	# counter of objects created.

my %attribute = (
	_n_objects => \$_n_objects,
	_log => get_logger("Visio"),
	extension => EXT_DOCUMENT,
	dir => ".",
	filename => undef,
	fho => \*STDOUT,
	_id => {	# see open()
		'Shape' => 1,
		'Master' => 0,
		'Page' => 0,
		'Window' => 0,
	},
	_loaded => 0,	# flag shows if a document is in-memory
	search_attr => [ ATTR_SEARCH ],
	_xml => undef,
);


#INIT { };


sub AUTOLOAD {
	my $self = shift;
	my $type = ref($self) or croak("self is not an object");

	my $name = $AUTOLOAD;
	$name =~ s/.*://;   # strip fully−qualified portion

	unless (exists $self->{_permitted}->{$name} ) {
		confess "no attribute [$name] in class [$type]";
	}

	if (@_) {
		return $self->{$name} = shift;
	} else {
		return $self->{$name};
	}
}


sub new {
	my ($class) = shift;
	my $self = { _permitted => \%attribute, %attribute };

	++ ${ $self->{_n_objects} };

	bless ($self, $class);

	$self->_log->debug(sprintf "created new object [%p]", $self);

	my %args = @_;	# start processing any parameters passed
	my ($method,$value);
	while (($method, $value) = each %args) {

		confess "SYNTAX new(method => value, ...) value not specified"
			unless (defined $value);

		$self->$method($value);
	}

	$self->{_xml} = new Visio::XML;

	$self->_log->debug(sprintf "_id [%p]", $self->_id);

	return $self;
}


sub list {
#	LIST directory contents for matching files!
	my $self = shift;
	my $dir = shift;

	if (defined $dir) {
		chdir($dir) || confess("chdir($dir) failed");
		$self->dir($dir);
	}

	my @files = glob('*' . $self->extension);

	$self->_log->debug(sprintf '@files [%s]', Dumper(\@files));

	return @files;
}


sub edit {
	my $self = shift;

	$self->docprop("TimeEdited", $self->timestamp);
}


sub empty {
	my $self = shift;

	$self->_log->logwarn("ERROR no document to save; have you done a blank() or open()?");

	return -1;
}


sub add_page {
	my $self = shift;
	my $name = shift;

	confess "SYNTAX add_page(name)" unless (defined $name);

	return $self->empty
		unless ($self->_loaded);

	$self->edit;

	my $type = 'Page'; 
	my ($top,undef) = $self->_xml->find($self->_xml->root, "$type" . "s");
	my ($page,undef) = $self->_xml->create($top, $type);

	$page->setAttribute('ID', ++$self->_id->{$type});
	$page->setAttribute('NameU', $name);

	$self->_log->debug(sprintf "filename [%s] _id [%p] _id [%s]", defined($self->filename) ? $self->filename : "(null)", $self->_id, Dumper($self->_id));
#	$self->_log->debug(sprintf "xmldoc [%s]", $self->_xml->doc);
#	$self->_log->debug(sprintf "xmlroot [%s]", $self->_xml->root);

	return $name;
}


sub docprop {
	my $self = shift;
	my $prop = shift;
	my $value = shift;

	confess "SYNTAX docprop(property, [value])" unless (defined $prop);

	return $self->empty
		unless ($self->_loaded);

	if (defined $value) {
		$self->_xml->set_property("DocumentProperties", $prop, $value);
	}

	return $value;
}


sub blank {
	my $self = shift;

	my $doc = $self->_xml->prepare(NN_DOCUMENT, NN_SECTIONS);

	$self->_loaded(1);

	$self->docprop("TimeCreated", $self->timestamp);

	return $doc;
}


sub open {
	my $self = shift;
	if (@_) { $self->filename(shift); }

	confess("SYNTAX open(filename)")
		unless defined($self->filename);

	confess(sprintf "file not found [%s]", $self->filename)
		unless (-f $self->filename);

	if ($self->_xml->read($self->filename) eq NN_DOCUMENT) {

		$self->_loaded(1);

		# observed business rules
		#	Shape ID starts at 1
		#	Master ID orders 0, 2, 3, ...
		#	Shape ID in master >= 4
		#	Page ID starts at 0
		#	Window ID starts at 0

		for (qw/ Master Page Window /) {
			my ($top,undef) = $self->_xml->find($self->_xml->root, $_ . 's');
			my (undef,$id) = $self->_xml->find($top, $_);

			$self->_id->{$_} = $id;
		}

		$self->_log->debug(sprintf "filename [%s] _id [%s]", $self->filename, Dumper($self->_id));

		return 0;
	}

	return -1;
}


sub save {
	my $self = shift;
	if (@_) { $self->filename(shift); }

	confess("SYNTAX save(filename)")
		unless defined($self->filename);

	return $self->empty
		unless ($self->_loaded);

	$self->docprop("TimeSaved", $self->timestamp);

	unless ($self->_xml->write($self->filename)) {

		return 0
			if (-f $self->filename);
	}

	$self->_log->warn(sprintf "file not written [%s]", $self->filename);

	return -1;
}


sub timestamp {
	my $self = shift;
	my $value = shift;	# in format expected by time()

	$value = time
		unless(defined $value);

	my $stamp = POSIX::strftime("%Y-%m-%dT%H:%M:%S", localtime($value));

	$self->_log->debug("returning [$stamp]");

	return $stamp;
}


DESTROY {
	my $self = shift;

	-- ${ $self->{_n_objects} };
};

#END { }

1;

__END__

=head1 VERSION

___EUMM_VERSION___

=head1 AUTHOR

Copyright (C) 2012  Tom McMeekin

=head1 SEE ALSO

L<perl>.

=cut



example/vsdx.pl
#!/usr/bin/perl

use warnings;
use strict;

# ref. https://learn.microsoft.com/en-us/office/client-developer/visio/how-to-manipulate-the-visio-file-format-programmatically
# ref. https://stackoverflow.com/questions/59935217/xmllibxml-issue-finding-xml-nodes-with-a-namespace

use Data::Dumper;
use OPC;
use XML::LibXML;
use XML::LibXML::PrettyPrint;

use constant NS_OPC => "http://schemas.openxmlformats.org/package/2006/relationships";

sub put_node {
	my $node = shift;

	printf "find [%s]  ", $node->getName;
	printf "nodeName [%s]  ", $node->nodeName;
	printf "localname [%s]  ", $node->localname;
#	printf "to_literal [%s]  ", $node->to_literal;

	my ($name, $value); for my $ns ($node->getNamespaces) {
		printf "name [%s]  ", $ns->name;
		printf "getPrefix [%s]  ", $ns->getPrefix;
		printf "value [%s]  ", $ns->value;

		$name = $ns->name;
		$value = $ns->value;

		last;
	}

	printf "\n";

	return ($name, $value);
}

my $pp = XML::LibXML::PrettyPrint->new(indent_string => "  ");

my $pnv = "Visio_harness.vsdx";

printf "opening [$pnv]\n";

my $doc = OPC->new($pnv);

printf "ref(doc) = [%s]\n", ref($doc);

#printf "root = [%s]\n", Dumper($doc->Root);

printf "IsBinary = [%s]\n", $doc->IsBinary;

printf "\n";
printf "PartNames\t  PartContentType\n";
#printf "PartNames [%s]\n", join(' ', $doc->PartNames);
my %pct; for my $pn ($doc->PartNames) {

	my $pt =  $doc->PartContentType($pn);

	printf "$pn\t  $pt\n";

	if (exists $pct{$pt}) {
		push @{ $pct{$pt} }, $pn;
	} else {
		$pct{$pt} = [ $pn ];
	}

	my $xml = $doc->PartContents($pn);
#	print "  PartContents [$xml]\n";
	my $dom = XML::LibXML->load_xml(string => $xml);
	$pp->pretty_print($dom);
	print $dom->toString;

#	for ($dom->findnodes('//@*')) {
#		printf "attr [%s]\n", $_->getName;
#	}

	for ($dom->findnodes('*')) {
		my ($nsn, $nsu) = put_node($_);

		my $this = $_->getName;
		printf "ELEMENT [$this]\n";

		my $xpc = XML::LibXML::XPathContext->new($dom);

		printf "\nregistering namespace [$nsn] for [$nsu]\n";
		$xpc->registerNs($nsn, $nsu);

		my $xp = sprintf "//$nsn:*"; printf "xp [$xp]\n";
		for ($xpc->findnodes($xp)) {
			put_node($_);

			for my $attr ($_->attributes) {

				my $atn = $attr->getName;
				my $atv = $attr->getValue;

				printf "attribute [$atn] = [$atv]\n";

				next unless ($this eq "Relationships");

				my $which = (); if ($atn eq 'Type') {
					$which = 'type';
				} elsif ($atn eq 'Id') {
					$which = 'id';
				} else {
					print "skipping irrelevant attribute\n";
				}
				if (defined $which) {

					print "polling relationships [$which]=[$atv]\n";

					for ($doc->Relations($pn, $which => $atv)) {
						printf "XXX $which relation [%s]\n", Dumper($_);
					}
				}
			}
		}
	}

	printf "\n";
}
printf "pct [%s]\n", Dumper(\%pct);

__END__

for ($doc->RelationNodesFromDoc('/visio/pages/_rels/pages.xml.rels')) {
	printf "RelationNodesFromDoc [%s]\n", Dumper($_);
}
for ($doc->RelationsByType('/visio/pages/_rels/pages.xml.rels')) {
	printf "RelationsByType [%s]\n", Dumper($_);
}


example/visio_parser.pl
#!/usr/bin/perl
#
# visio_parser.pl - summarise elements and attributes for a Visio XML diagram
# Revision ___EUMM_VERSION___, Copyright (C) 2014 Thomas McMeekin
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License,
# or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#
=head1 NAME

visio_parser.pl - summarise elements and attributes for a Visio XML diagram

=head1 SYNOPSIS

perl -I ../lib -I ../../Visio-XML/lib -w visio_parser.pl 

[-h, --help]
[-m, --manual]
file ...

=head1 OPTIONS

=over 8

=item B<--help>

Prints a brief help message and exits.

=item B<--manual>

Prints the manual page and exits.

=back

=cut

use strict;

use Carp qw/ croak carp /; 
use Data::Dumper;

use Cwd;
use File::Copy;;
use Logfer qw/ :all /;

use Visio;
use XML::LibXML qw/ :libxml /;


# ---- global variables ----
my $g_log = get_logger(__FILE__);
my %g_xml;


# ---- initialisation ----
#printf "argv [%s]\n", scalar @ARGV;
die "usage: perl -I ../lib -I ../../Visio-XML/lib -w visio_parser.pl file ...\n" if (@ARGV < 1);
my $vdx = Visio->new();


# ---- sub-routines ----
sub blank {
	my ($value)=@_;
	my $blank = "(blank)";

	return $blank unless (defined $value);

	$g_log->debug("before value [$value]");

	$value =~ s/^\s*$//g;

	$g_log->debug(" after value [$value]");

	return $blank
		if ($value eq '');

	return $value;
}


sub offset {
	my $level = shift;

	my $tab = "";

	while (--$level >= 0) {

		$tab .= "    ";
	}

	return $tab;
}


sub record {
	my ($key, $value)=@_;

	chomp $value;

	if (exists $g_xml{$key}) {

		if (exists $g_xml{$key}->{$value}) {

			$g_xml{$key}->{$value}++;
		} else {
			$g_xml{$key}->{$value} = 1;
		}
	} else {

		$g_xml{$key} = { $value => 1 };
	}
}


sub whatami {
	my ($xpath,$node,$level)=@_;

	return -1 unless defined ($node);

	my $type = $node->nodeType;
	my $name = blank($node->nodeName);
	my $s_type;
	my $value = "";

	if ($type == XML_ELEMENT_NODE) {
		$s_type = "element";
		$value = blank($node->nodeValue);

	} elsif ($type == XML_ATTRIBUTE_NODE) {
		$s_type = "attr";
		$value = blank($node->getValue);

		my $key = join('@', $xpath, $name);
		record($key, $value);

	} elsif ($type == XML_TEXT_NODE) {
		$s_type = "text";
		$value = blank($node->nodeValue);

		$value = 'XXXXX' if ($xpath =~ /\/PreviewPicture|\/Icon/);

		record($xpath, $value);

	} elsif ($type == XML_COMMENT_NODE) {

		$s_type = "comment";

	} else {
		$s_type = "other";
	}
	$g_log->debug(sprintf "%stype '%s(%s)' path '%s' name '%s' %s", offset($level), $s_type, $type, $xpath, $name, ($value eq "") ? "" : sprintf('value [%s]', $value));

	return $type;
}


sub recurse {
	my ($parent,$level,$path)=@_;

	my $xpath = join('/', $path, $parent->nodeName);

	$g_log->debug("xpath [$path]");

	whatami($xpath, $parent, $level);

	for ($parent->attributes()) {

		whatami($xpath, $_, $level);

		#$g_log->debug(sprintf "%s type [%s] attr [%s] value [%s]", offset($level), $_->nodeType, $_->nodeName, blank($_->getValue));
	}

	my $children = 0;
	for ($parent->childNodes) {
		$g_log->debug(sprintf("%schild [%s]", offset($level), $children++));
		recurse($_, $level+1, $xpath);
	}
}


# ---- main ----
for my $m_fn (@ARGV) {

	next if ($vdx->open($m_fn));

	#$g_log->debug(sprintf 'XML [%s]', Dumper($vdx->_xml));


	#$g_log->debug(sprintf 'nodes [%s]', Dumper(\@nodes));

	recurse($vdx->_xml->root, 0, "");
}

# dump values
$g_log->info("listing of values");
$g_log->info("path\texample value(s)");

for my $xpath (sort(keys %g_xml)) {

	$g_log->info(sprintf "%s:\t%s",
		$xpath,
		join(', ', sort(keys %{ $g_xml{$xpath} }))
	);
}


__END__

=head1 DESCRIPTION

Parse Visio XML diagram and summarise elements and attributes.

=head1 VERSION

___EUMM_VERSION___

=head1 AUTHOR

B<Tom McMeekin> tmcmeeki@cpan.org

This code is distributed under the same terms as Perl.

=head1 SEE ALSO

L<perl>.

=cut

